test/end-to-end/offer.js

```diff
@@ -298,6 +299,139 @@ describe("Offer endpoint tests", () => {
             });
         });
 
+        describe("Before reaching the offers limit while having past offers", () => {
+            const testOffers = [];
+            for (let i = 1; i < CompanyConstants.offers.max_number; ++i) {
+                testOffers.push({
+                    title: `Active Test Offer no${i}`,
+                    publishDate: new Date(Date.now() - (DAY_TO_MS)),
+                    publishEndDate: new Date(Date.now() + (DAY_TO_MS)),
+                    description: "For Testing Purposes",
+                    contacts: ["geral@niaefeup.pt", "229417766"],
+                    jobType: "SUMMER INTERNSHIP",
+                    fields: ["DEVOPS", "MACHINE LEARNING", "OTHER"],
+                    technologies: ["React", "CSS"],
+                    // owner: will be set in beforeAll
+                    // ownerName: will be set in beforeAll
+                    location: "New Testing Avenue, New Testing, 234",
+                });
+            }
+
+            testOffers.push({
+                title: "Past Test Offer",
+                publishDate: new Date(Date.now() - (2 * (DAY_TO_MS))),
+                publishEndDate: new Date(Date.now() - (DAY_TO_MS)),
+                description: "For Testing Purposes",
+                contacts: ["geral@niaefeup.pt", "229417766"],
+                jobType: "SUMMER INTERNSHIP",
+                fields: ["DEVOPS", "MACHINE LEARNING", "OTHER"],
+                technologies: ["React", "CSS"],
+                // owner: will be set in beforeAll
+                // ownerName: will be set in beforeAll
+                location: "New Testing Avenue, New Testing, 234",
+            });
```

Comment:
Create a function (can be inside the describe since it will only be used here) that receives the publish and publish end dates and returns the offer object, so that you don't need to repeat it multiple times.

```js
const generateTestOffer = (publishDate, publishEndDate) => ({
                title: "Past Test Offer",
                publishDate,
                publishEndDate,
                description: "For Testing Purposes",
                contacts: ["geral@niaefeup.pt", "229417766"],
                jobType: "SUMMER INTERNSHIP",
                fields: ["DEVOPS", "MACHINE LEARNING", "OTHER"],
                technologies: ["React", "CSS"],
                // owner: will be set in beforeAll
                // ownerName: will be set in beforeAll
                location: "New Testing Avenue, New Testing, 234",
})
```

You can then use it like this: 
```js
Array(CompanyConstants.offers.max_number - 1)
     .fill(generateTestOffer(
          new Date(Date.now() - (DAY_TO_MS)),
          new Date(Date.now() + (DAY_TO_MS))
      ));
```

Notice how easier it is now to know that you are specifically creating `max offers - 1` :)

you can later `testOffers.push(generateTestOffer(other values))`

Reply:
I decided to create the generate function in the scope of the "Offer endpoint tests" describe since it can be useful for future tests. You'll see it in detail when I finish up but does it sound good to you?

Reply:
In that case you can replace the `offer` variable on the top and if you slightly modify the generator to allow undefined args, something like: 
```js
(publishDate, publishEndDate)=> ({
.........
publishDate: publishDate || new Date(Date.now() - (DAY_TO_MS)),
publishEndDate: publishEndDate || new Date(Date.now() + (DAY_TO_MS)),
........
```

Then you can just replace the usages of the offer variable in all other tests with generateTestOffer() and do the full refactor, which shouldn't take too much work and will clean up everything a bit